docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+
+### Failures and Recovery Strategy
+
+In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption.
+
+#### Types of Failures During Update
+
+##### Drain Timeout/Failure
+
+- A node is marked as `ready-for-update` only after a successful drain.
+- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
+- If a drain operation times out or fails, the node will be forcefully drained.
+
+##### Kubernetes Update Failures
+
+- **Rollback on Failure:** If a Kubernetes version update fails, `gardener-node-agent` will roll back the `kubelet` to the version specified in the last applied `OperatingSystemConfig`. This ensures that the node remains operational.
+  - **Update Halt on Failure:** `gardener-node-agent` will label the node with an `update-failed` label. This label signals `machine-controller-manager` to stop further updates.
+
+##### OS Update Failures
+
+- **Before rebooting for update** - Update cannot start if the machine image version is unavailable, network issues prevent the image pull, or disk storage is insufficient (although proper garbage collection should prevent this); classified as retriable or non-retriable.
+- **Failed to boot the new OS** - If the node fails to boot into the new OS version and reverts to the previous version, `gardener-node-agent` will annotate it to mark an update attempt, but the root cause may remain unlogged as filesystems mount later in the booting process; classified as non-retriable.
+- **Total failure** - In cases where the system fails to boot into either of the OS versions, the update attempt will timeout and be marked as failed; classified as non-retriable.
+
+![FailureStrategy](assets/inplace-failurestrategy.png "InPlace Failure State Diagram")
+
+#### Roles in Failure Handling
+
+**`gardener-node-agent`**
+`gardener-node-agent` is responsible for initiating updates, with specific timeouts set for Kubernetes and machine image updates. When an error occurs it will be classified as follows:
+
+- **Retriable Error**: Temporary errors, such as network interruptions, prompt `gardener-node-agent` to retry the update after a delay. Retriable error will also get classified as non-retriable error after a certain number of retries and `gardener-node-agent` labels the node with an `update-failed` label.
+- **Non-Retriable Error**: For critical issues, such as incompatible OS versions, `gardener-node-agent` labels the node with an `update-failed` label.
```

Comment:
If we want to make `gardener-node-agent` classify operating system issues, we assume that `gardener-node-agent` will get running after big issues, e.g. a corrupt operating system image. Is that assumption valid? Should this be stated explicitly?

Reply:
Only the error when the OS update trigger is not possible will get classified, like version not found, unable to pull the new image and so and so.
For the case where the OS fails to boot/crashes, we will anyway not get the errors of why the OS update failed, and GNA will not be able to come up, in this case, MCM will timeout, and consider it a failed update.